apiserver/facades/agent/instancemutater/instancemutater.go

```diff
@@ -264,22 +272,33 @@ func (api *InstanceMutaterAPI) getMachine(canAccess common.AuthFunc, tag names.M
 	return machine, nil
 }
 
-func (api *InstanceMutaterAPI) machineLXDProfileInfo(m Machine, unitNames []string) ([]params.ProfileChangeResult, []string, bool, error) {
-	if instId, err := m.InstanceId(); err != nil && params.IsCodeNotProvisioned(err) {
+// lxdProfileInfo holds the profile information for the machineLXDProfileInfo
+// to provide context to the result of the call.
+type lxdProfileInfo struct {
+	InstanceId      instance.Id
+	MachineProfiles []string
+	ProfileChanges  []params.ProfileChangeResult
+}
+
+func (api *InstanceMutaterAPI) machineLXDProfileInfo(m Machine, unitNames []string) (lxdProfileInfo, bool, error) {
+	var empty lxdProfileInfo
+
+	instId, err := m.InstanceId()
+	if err != nil && params.IsCodeNotProvisioned(err) {
 		// There is nothing we can do with this machine at this point. The
 		// profiles will be applied when the machine is provisioned.
 		logger.Tracef("Attempting to apply a profile to a machine that isn't provisioned %q", instId)
-		return nil, nil, false, nil
+		return empty, false, nil
 	}
 	model, err := api.st.Model()
 	if err != nil {
-		return nil, nil, false, errors.Trace(err)
+		return empty, false, errors.Trace(err)
 	}
 	modelName := model.Name()
 
 	machineProfiles, err := m.CharmProfiles()
 	if err != nil {
-		return nil, machineProfiles, false, errors.Trace(err)
+		return empty, false, errors.Trace(err)
 	}
 	changeResults := make([]params.ProfileChangeResult, len(unitNames))
```

Comment:
You can do a machine.Units() from the model cache now.

The big difference with model cache vs state.  Is that the individual structs don't give you the thing, you have to get it from the model.  For instance, you get the application name from the model cache unit.  Then ask the model cache model for the application based on the name.